---
title: Brain Impairment; Resulting in Reduction of Expression of Emotion, Motivation,
  and Goal-Directed Behavior in Heroin Users.
author: "Narad Mishra"
date: "December 15, 2017"
output: html_document
--- 

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(clusterProfiler)
library(Biobase)
library(limma)
library(GEOquery)
library(affy) 
library(arrayQualityMetrics)
library(AnnotationDbi)
library(topGO)
library(devtools)
library(sva)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(pander)
library(scatterplot3d)
library(KEGGREST)
library(DT)
```



## Introduction

	Importance of brain cells in regulating human activities are known to everyone. Nucleus accumbens, a region in the basal forebrain rostral to the preoptic area of the hypothalamus is responsible for regulating our emotion expression, motivation, and goal-directed behavior. These cells, if damaged or if affected by chemicals will have deteriorating effect in human behavior. Harmful drugs, such as cocaine, heroin, and other opioids are linked to having such damage on nucleus accumbens cells, which cause the individual to behave different than normal people. This study compares the gene expression of heroin users with the normal population. The study aims to find at least some genes differentially expressed in heroin users compared to the non-user population. Similar comparison has been done before with the data I am going to use (Hurd Yl, et.al. 2016).
	
## Method

	For this study, genes expression profile of human brain cells from heroin users and normal population were used.The data was obtained from NCBI public dataset.The geo-accession Id for the data is GSE87823 and platform GPL96 [HG-U133A]. It is an expression profiling by array, by Affymetrix Human Genome U133A array. The data was first downloaded directly from GEOdataset, using getGEO function from *Bioconductor*. The data set was visually accessed using box plot of the arrays and distribution of the intensities for quality control measures.Since, most (not all) of the experimental data set in GEO-datasets are already normalized, rawdata for the experiment was also loaded with *ReadAffy* function from *Bioconductor*. Box-plots and distribution of the intensities for the data read with *ReadAffy* was then accessed and the result was compared with the first box plots. Then Robust Multi-Array Average (RMA) method from *affymatrix* was applied to the expression set for normalization of the data. The RMA performs normalization by quantiles on the intensities of the probes and performs expression summarization using the Median Polish method. RMA is a log scale linear additive model that does not use mismatch probes (MM).Even though RMA decreases accuracy, other method don't necessarily performs any better. While it decreases the accuracy, it increases precion. Surrogate Variable Analysis (SVA) was performed to eliminate unknown batch and *ComBat* were performed to remove the batch effect known covariate (case vs control). The outcome of the two methods was graphically assessed. SVA works by identifying the number of latent factor that it needs to estimate. Then the Principal Components Analysis (PCA) was applied to both rawdata and the data with corrected batch effect, to further make sure of the batch effect. Hierarchical clustering of top ten percent of highly variable genes were clustered using 'Average Linkage', and 'Complete Linkage'. Clustering was performed on the data, both before and after adjusting for batch effect. Filtering procedure was performed using the *nsFilter* function to reduce the number of comparisons and increase power.  Features without an Entrez Gene ID annotation or mapping the same gene, without annotation to at least one GO term (BP, CC or MF) ontology, genes with an interquartile range (IQR) in the lower 50% of the IQR distribution, and Affymetrix control probes were deleted. Raw p-values were adjusted for multiple hypothesis testing using the false discovery rate (FDR) method of Benjamini and Hochburg. To find the significantly differentially expressed genes, *Limma* method in *Bioconductor* was used. Limma is a linear model based gene analysis package that provides ability to analyse comparisons between multiple gene sets.Top 10 most differentially expressed subsetted from the list of differentially expressed genes and were box plotted. Then functional enrichment analysis was performed with three different methods. First gene enrichment analysis was performed using Fisher test, then the enrichment was performed for Gene Ontology (GO), and enrichment analysis under KEGG pathways was performed.  


## Result

	The data consist of 49 samples, where 27 of the samples are from control group which are normal Caucasian population and 22 heroin users. The data consisted of 22283 gene sets. The box-plot obtained from the rawdata loaded using *getGEO* looked very much normalized. Histogram of the intensity for a probe plotted also looks normally distributed. Box-plot of the data loaded using *ReadAffy* confirmed that the initially loaded data was normalize. The two plots and the histograms are shown in Figure 1, and Figure 2. Since the data for this study was already normalized, quality evaluation were satisfactory and the data didn't need any cleaning. If the quality check had suggested some cleaning, the *arrayQualityMetrics* function from *Bioconductor* would be used to control for the quality issue. Box-plot of the normalized data, of the data read using *ReadAffy* was almost identical to the boxplot of data read using *getGEO*, which furthur confirms that the data used for this study was already normalize and summarized (fig 3). Batch effect correction methods used in the study seems to remove the batch present, as the accessed box-plots and histogram are very similar between two groups when batch was known (fig 4). The histogram of a probe from batch corrected between SVA and ComBat also looks very similar. And the plot of the coefficients of SVA versus ComBat is linear with the slope of almost 1. Principal Component Analysis performed on both the raw data and the batch corrected data didn't show any difference in distribution among the two components (fig 5). Cluster maps are presented from clustering top 10% of most highly variable genes (fig 6). The filter function was able to filter out 16825 probe ids that didn't have one or more of specified filteration above. Using FDR as the multiple testing adjustment, limma was able to find 379 differentially expressed genes.Fisher test for functional enrichment showed that chromosome 2, 3, 7, 8, 11, 15, 17, 18, 20, 21, 22, and Y were enriched, and chromosome 1,4,5,9,10,12,13,14,16,19, and X are depleted. Gene was called Enriched if odds ratio was greater than 1 and depletd if the odds ratio is less than one. Enrichment analysis performed using Gene Ontology yielded similar results as the fisher test. The output figure for the gene ontology analysis is shown in Figure 7. KEGG pathway analysis gave the description of physiological pathways and some diseases like cancer, Huntington's disease, etc. that the enriched genes are associated with. 
	
	
Table :1 Top ten differentially expressed genes
```{r,warning=FALSE,message=FALSE,echo=FALSE} 
mydata<-getGEO('GSE87823', GSEMatrix = TRUE, AnnotGPL = TRUE, destdir = "C:/Users/Narad Mishra/Desktop/project/new")
mydata<-getGEO(filename = "C:/Users/Narad Mishra/Desktop/project/new/GSE87823_series_matrix.txt.gz", destdir = ".")

edata<-exprs(mydata)      #expression matrix from series data
pheno<-pData(mydata)      #pheno data 
#the above phenodata will be used to get information for batch effect.

library(hgu133a.db)
annotation(mydata)<-"hgu133a.db"

f.edata <- nsFilter(mydata, require.entrez = TRUE, require.GOBP = TRUE, 
    require.GOCC = TRUE, require.GOMF = TRUE, require.CytoBand = FALSE, 
    remove.dupEntrez = TRUE, var.func = IQR, var.cutoff = 0.5, var.filter = TRUE, 
    filterByQuantile = TRUE, feature.exclude = "^AFFX")

f.edata$filter.log
filteredset <- f.edata$eset  # The new ExpressionSet
# group names for all samples in a series
#partial code taken from Geo2R
gsms <- "1111111111111111111111111110000000000000000000000"    # 1 being control and 0 being heroin users
sml <- c()
for (i in 1:nchar(gsms)) { sml[i] <- substr(gsms,i,i) }
sml <- paste("G", sml, sep="")

# set up the data and proceed with analysis
sml <- paste("G", sml, sep="")    # set group names
fl <- as.factor(sml)
filteredset$description <- fl
design <- model.matrix(~ description + 0, filteredset)
colnames(design) <- levels(fl)
fit <- lmFit(filteredset, design)
cont.matrix <- makeContrasts(GG1-GG0, levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2)
tT <- topTable(fit2, n=Inf, adjust.method ="fdr", p.value=0.05)
datatable(tT)
tT[order(tT$adj.P.Val),]
top10<-tT[1:10,]

```


Fig 1: Boxplot of the samples and histogram of a sample from the data loaded with getGEO
```{r, warning=FALSE, message=FALSE, echo=FALSE}

##Quality Control
library("RColorBrewer")
usr.col=brewer.pal(9, "Set1")
mycols=rep(usr.col,each=3)
boxplot(log2(edata), col=mycols, las=3, cex.axis=0.5, names=colnames(edata))
hist(log((edata)[,1],2))

```

Fig 2: Boxplot of the samples and histogram of a distribution of samples from the data loaded with ReadAffy
```{r,warning=FALSE, message=FALSE, echo=FALSE}
files<-list.celfiles() #provided the cel.files are present in working directory
datDir <- "C:/Users/Narad Mishra/Desktop/project/new"
files<-list.celfiles(path=datDir) #provided the cel.files are present in working directory
affy1<-ReadAffy(filenames = files,celfile.path=datDir)

hist(affy1, lty=rep(1,49), col=mycols)
legend("topright",rownames(pData(affy1)),lty=rep(1,49), col=mycols, cex=0.6)

#accessing the boxplots
boxplot(affy1, col=mycols, las=3, cex.axis=0.5, names=colnames(edata))
```


Fig 3:Boxplot of the samples from normalization of the data loaded with ReadAffy
```{r,warning=FALSE, message=FALSE, echo=FALSE}
eset<-rma(affy1)
#this performs background correction, quantile normalize, and calculates expression set (summarize using median polish)
expset<-exprs(eset) #get the expression estimates for each array
#take a look at few result
expset[1:5,1:5]
#accessing the boxplot after normalization. (look the same as the plot for mydata)
boxplot(expset, col=mycols, las=3, cex.axis=0.5, names=colnames(expset))

```

Fig 4:Box-plot for control and cases by SVA

```{r,warning=FALSE, message=FALSE, echo=FALSE}

batch=pheno$source_name_ch1
modcombat <- model.matrix(~1, data = pheno)
mod <- model.matrix(~source_name_ch1, data = pheno)

combat_edata <- ComBat(dat = edata, mod = modcombat, batch=batch,par.prior = TRUE, 
                       prior.plots = FALSE)

combat_fit <- lm.fit(mod, t(combat_edata))
combat_fit <- lmFit(combat_edata, mod)

#Applying SVA method 
sva1 <- sva(edata, mod , modcombat, n.sv = 2)
summary(lm(sva1$sv ~ pheno$source_name_ch1))
boxplot(sva1$sv[, 2] ~ pheno$source_name_ch1)
```


Fig 5:Histogram of coefficients from ComBat and SVA method
```{r,warning=FALSE, message=FALSE, echo=FALSE}
modsv <- cbind(mod, sva1$sv)
fitsv <- lmFit(edata, modsv)

par(mfrow=c(2,1))
hist(combat_fit$coefficients[, 2], col = 2, breaks = 100) # hisogram applying combat method
hist(fitsv$coefficients[, 2], col = 2, breaks = 100)
```


Fig 6: Plot of Coefficient of SVA and ComBat 
```{r,warning=FALSE, message=FALSE, echo=FALSE}
par(mfrow = c(1, 1))
plot(fitsv$coefficients[, 2], combat_fit$coefficients[, 2], col = 2, xlab = "SVA", 
     ylab = "Combat", xlim = c(-5, 5), ylim = c(-5, 5))
abline(c(0, 1), col = 1, lwd = 3)
```


Fig 7: Comparison of three components of PCA before batch adjustment
```{r,warning=FALSE, message=FALSE, echo=FALSE}
options(stringsAsFactors = FALSE)

sourcetype<-pheno$source_name_ch1   # info on whether control or heroin users
#edata<-exprs(mydata) from above
pca.data<-edata
# rename samples
grps <- c(rep("Cont",27), rep("User",22))
grpcol <- c(rep("blue",27), rep("red",22))
colnames(pca.data) <- paste(grps, colnames(pca.data), sep="-")
 
# remove NAs
pca.data <- na.omit(as.matrix(pca.data))
 
# transpose
pca.data <- t(pca.data)
 
# inspect pca.data
dim(pca.data)
pca.data[1:2,1:2]   #this should show control group as rowname
pca.data[30:31,1:2] #this should show user group
 
# compute PCA
pca <- prcomp(pca.data, scale=TRUE)
 
# identify variance in components
summary(pca)
 
# the first 2 component group 64% of the total variance
# the first 3 component group 79% of the total variance
# the first 4 component group 90% of the total variance
par(mfrow=c(3,1))
# components #1 and #2
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1&2", type="p", pch=10, col=grpcol)

# show other component pairs for the example
# components #1 and #3
plot(pca$x[,1], pca$x[,3], xlab="PCA1", ylab="PCA3", main="PCA for components 1&3", type="p", pch=10, col=grpcol)

# components #2 and #3
plot(pca$x[,2], pca$x[,3], xlab="PCA2", ylab="PCA3", main="PCA for components 2&3", type="p", pch=10, col=grpcol)
```

Fig 8: Comparison of three components of PCA after batch adjustment

```{r}

#PCA after batch correction.
x <- t(combat_edata)
pcr <- prcomp(x)

plot(pcr$x[,1], pcr$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1&2 after batch correction", type="p", pch=10, col=grpcol)

# show other component pairs for the example
# components #1 and #3
plot(pcr$x[,1], pcr$x[,3], xlab="PCA1", ylab="PCA3", main="PCA for components 1&3 after batch correction", type="p", pch=10, col=grpcol)

# components #2 and #3
plot(pcr$x[,2], pcr$x[,3], xlab="PCA2", ylab="PCA3", main="PCA for components 2&3 after batch correction", type="p", pch=10, col=grpcol)
```


Fig 9: Clustter Pattern before and after adjusting for batch using average linkage.
```{r}
 # Identifying 10% of more variable genes
funt <- apply(edata, 1, sd)
cut <- quantile(funt, 0.9)
list.var <- which(funt > cut)

top10 <- edata[list.var, ]   #before batch correction
top10combat<- combat_edata[list.var, ] #after batch correction

# average clustering
hclusbefore <- hclust(dist(top10), method = "average")
hclusafter <- hclust(dist(top10combat), method = "average")
```


Fig 10: Clustter Pattern before and after adjusting for batch using complete linkage.
```{r, message=FALSE, warning=FALSE, echo=FALSE}
par(mfrow = c(2, 1), mar = c(1, 4, 2, 1))
plot(hclusbefore, labels = FALSE, hang = -1, main = "before batch effect")
plot(hclusafter, labels = FALSE, hang = -1, main = "after batch effect")


# complete
hclus.bef <- hclust(dist(top10), method = "complete")
hclus.aft <- hclust(dist(top10combat), method = "complete")


par(mfrow = c(2, 1), mar = c(1, 4, 2, 1))
plot(hclus.bef, labels = FALSE, hang = -1, main = "before batch effect")
plot(hclus.aft, labels = FALSE, hang = -1, main = "after batch effect")
```

Fig 11:Box plot of top ten differentially expressed gene
```{r, echo=FALSE}
gsms <- "1111111111111111111111111110000000000000000000000"    # 1 being control and 0 being heroin users
sml <- c()
for (i in 1:nchar(gsms)) { sml[i] <- substr(gsms,i,i) }
sml <- paste("G", sml, sep="")

# set up the data and proceed with analysis
sml <- paste("G", sml, sep="")    # set group names
fl <- as.factor(sml)
filteredset$description <- fl
design <- model.matrix(~ description + 0, filteredset)
colnames(design) <- levels(fl)
fit <- lmFit(filteredset, design)
cont.matrix <- makeContrasts(GG1-GG0, levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2)
tT <- topTable(fit2, n=Inf, adjust.method ="fdr", p.value=0.05)
datatable(tT)
tT[order(tT$adj.P.Val),]
top10<-tT[1:10,]

```

```{r}
vars <- c("ID","adj.P.Val","P.Value","t","logFC","Gene.Symbol","Gene.Title")
topten<-top10[,colnames(top10) %in% vars]

# Matching the Id with expression matrix
edata1<-data.frame(edata)
edata1$probeid<-rownames(edata1)
top10gene<-semi_join(edata1, topten, by=c("probeid"="ID"))
rownames(top10gene)<-top10gene$probeid
top10gene$probeid<-NULL

# Boxplot of ten most differentially expressed gene
boxplot(top10gene, col=mycols, las=3, cex.axis=0.5, names=colnames(expset))
```
```



```{r,echo=FALSE}

allgene<-as.list(hgu133aSYMBOL)   ## Get full mapping table for probes to gene names

gene <- unlist(allgene[row.names(tT)])
gene<-data.frame(gene)
gene$probeid<-rownames(gene)
rownames(gene)<-NULL

chrloc <- as.list(hgu133aCHR)
chrloc <- unlist(chrloc)
chrloc <- data.frame(chrloc)
chrloc$Gene_ID <- rownames(chrloc) 
rownames(chrloc) <- NULL
chrloc<-chrloc[complete.cases(chrloc),]

names(chrloc)[1] <- "Chromosome"
chrloc$Chromosome <- as.character(chrloc$Chromosome)
chrloc2 <- inner_join(chrloc,gene, by=c("Gene_ID"="probeid"))
chrloc2$Gene_ID <- as.character(chrloc2$Gene_ID)

express.nn <- exprs(filteredset)
express.nn<-data.frame(express.nn)

express.nn$Gene_names <- rownames(express.nn)
rownames(express.nn) <- NULL
express3 <- left_join(express.nn, chrloc, by=c("Gene_names"="Gene_ID"))
express3$diff_exprs <- ifelse(express3$Gene_names %in% rownames(tT), 1, 0)
express3$diff_exprs <- factor(express3$diff_exprs, levels = c(1,0))

gene_chr <- data.frame(Gene_names = express3$Gene_names, Chromosome = express3$Chromosome, DiffExp = express3$diff_exprs)
gene_chr<-gene_chr[complete.cases(gene_chr),]
head(gene_chr)

mat <- matrix(ncol=length(unique(gene_chr$Chromosome)), nrow=dim(gene_chr)[1])
chrval <- c(1:22,"X","Y")
for(i in 1:24){
	mat[,i] <- ifelse(gene_chr$Chromosome==chrval[i],1,0)
	}
chrominfo <- data.frame(mat)
names(chrominfo) <- paste(rep("chr",24),c(1:22,"X","Y"),sep="")
chrominfo[] <- lapply(chrominfo, factor, 
              levels=c(1,0))
	
pval <- numeric()
or <- numeric() 
for(i in 1:24){
ft <- fisher.test(table(gene_chr$DiffExp,mat[,i]))
pval[i] <- ft$p.value
or[i] <- ft$estimate
}


# Enrichment analysis for Gene Ontology (GO)

# Enrichment analysis for Gene Ontology (GO)

sel.fn <- function(p.vals) { p.vals < 0.05 } # Returns a subset of most significant probes
affyLib <- paste(annotation(filteredset), "db", sep=".")
p.values <- tT$adj.P.Val
names(p.values) <- rownames(tT)

GOdata <- new("topGOdata",
               ontology = "BP", allGenes = p.values, geneSel = sel.fn,
               nodeSize = 10, # search GO terms with >= 10 genes
               annot = annFUN.db, affyLib = "hgu133a.db")
 
resultFisher <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
resultFisher

showSigOfNodes(GOdata, score(resultFisher), firstSigNodes = 2, useInfo = "all")
```


## Discussion

  Heroin overdose or excess use can be detrimental to mental health. In this study, heroin users and control groups microarray data was compared to see if there were any differentially expressed genes. The study was able to find some very specific genes and chromosomes that are differentially expressed. As we saw from different test such as principal component analysis both before and after the batch correction didn't show any divergence between the components, it confirms that the data didn't have any batch to control for. The linear plot of the coefficient of SVA method and ComBat also indicates that there was no any batch, so they both worked to remove the same amount of variation due to batch (unknown). From the the fisher test for functional analysis, it seems like at least one gene from each chromosome are either enriched or depleted. For more accurate inference, this study needs to compare more adjustment method for multiple testing and also look into performing may be somewhat more conservative approaches.

##References
 
 * Chapman. A, and Book. H.(2012) "Statistics and Data Analysis for Microarrays Using R and Bioconductor", CRC press, Taylor$Francis Group.
 
